tests/test_login.py

```diff
@@ -19,63 +17,63 @@ def test_login_writes_input_key_to_netrc_file(self):
         scrapy_cfg_with_username = """
 [deploy]
 username = KEY_SUGGESTION
         """
-
-        with self.runner.isolated_filesystem():
-            with open('scrapy.cfg', 'w') as f:
-                f.write(scrapy_cfg_with_username)
-
-            result = self.runner.invoke(login.cli, input='123')
-            err = 'Unexpected output: %s' % result.output
-            self.assertTrue('KEY_SUGGESTION' in result.output, err)
+        result = self._run(files={'scrapy.cfg': scrapy_cfg_with_username}, read_scrapy_cfg=True)
+        err = 'Unexpected output: %s' % result.output
+        self.assertTrue('KEY_SUGGESTION' in result.output, err)
 
     def test_login_suggests_shub_apikey_as_key(self):
-        with self.runner.isolated_filesystem():
-            with patch.object(login, '_read_scrapy_cfg_key', return_value=None):
-                result = self.runner.invoke(login.cli, input='123',
-                                            env={'SHUB_APIKEY': 'SHUB_APIKEY_VALUE'})
-
-                err = 'Unexpected output: %s' % result.output
-                self.assertTrue('SHUB_APIKEY_VALUE' in result.output, err)
+        result = self._run(env={'SHUB_APIKEY': 'SHUB_APIKEY_VALUE'})
+        err = 'Unexpected output: %s' % result.output
+        self.assertTrue('SHUB_APIKEY_VALUE' in result.output, err)
 
     def test_login_can_handle_invalid_scrapy_cfg(self):
-        # given
-        invalid_scrapy_cfg = 'invalid content'
+        result = self._run(files={'scrapy.cfg': 'invalid content'})
+        self.assertEqual(0, result.exit_code, result.exception)
 
+    def test_login_attempt_after_login_doesnt_lead_to_an_error(self):
         with self.runner.isolated_filesystem() as fs:
-            login.auth.NETRC_FILE = os.path.join(fs, '.netrc')
+            # when
+            self._run(fs=fs)
+            result = self._run(fs=fs)
 
-            with open('scrapy.cfg', 'w') as f:
-                f.write(invalid_scrapy_cfg)
+            # then
+            self.assertEqual(0, result.exit_code)
+            self.assertTrue('already logged in' in result.output, result.output)
 
-            with patch.object(login, '_is_valid_apikey', return_value=True):
-                # when
-                result = self.runner.invoke(login.cli, input=self.VALID_KEY)
+    def _run(self, user_input=VALID_KEY, files=None, fs=None, env=None, read_scrapy_cfg=False):
+        """Invokes the login cli on an isolated filesystem"""
 
-                # then
-                self.assertEqual(0, result.exit_code, result.exception)
+        def write_local_test_files():
+            for path, content in (files or {}).iteritems():
+                with open(path, 'w') as f:
+                    f.write(content)
 
-    def test_login_attempt_after_login_doesnt_lead_to_an_error(self):
-        with self.runner.isolated_filesystem() as fs:
+        def invoke():
+            return self.runner.invoke(login.cli, input=user_input, env=env)
+
+        def run():
```

Comment:
The problem is that the login command is just code that reads/writes files, input and makes network requests basically, there isn't much logic. 

We could try a different design for the login, more OO, for example. But I'm not sure it's worth the trouble right now as the feature is already pretty stable.

At least in the refactored code the test purpose is more obvious, IMO.